File: dfs/110.balanced-binary-tree.py
from typing import Optional

# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        l = self.getDepth(root.left)
        if l == -1:
            return -1
        r = self.getDepth(root.right)
        if abs(l - r) > 1 or r == -1:
            return -1
        return max(l, r) + 1
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return self.getDepth(root) != -1

# @leet end

File: DFS/110.balanced-binary-tree.go
package leetcode

// @leet start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func getDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	l := getDepth(root.Left)
	if l == -1 {
		return -1
	}
	r := getDepth(root.Right)
	if r == -1 || abs(l-r) > 1 {
		return -1
	}
	return max(l, r) + 1
}


func isBalanced(root *TreeNode) bool {
	return getDepth(root) != -1
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

// @leet end

